keel-web/src/test/kotlin/com/netflix/spinnaker/keel/ApplicationServiceTests.kt

```diff
@@ -0,0 +1,358 @@
+import com.netflix.spinnaker.keel.core.api.PipelineConstraint
+import com.netflix.spinnaker.keel.core.api.StatefulConstraintSummary
+import com.netflix.spinnaker.keel.ec2.SPINNAKER_EC2_API_V1
+import com.netflix.spinnaker.keel.events.ResourceValid
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.HAPPY
+import com.netflix.spinnaker.keel.persistence.memory.InMemoryArtifactRepository
+import com.netflix.spinnaker.keel.persistence.memory.InMemoryDeliveryConfigRepository
+import com.netflix.spinnaker.keel.persistence.memory.InMemoryResourceRepository
+import com.netflix.spinnaker.keel.services.ApplicationService
+import com.netflix.spinnaker.keel.test.combinedInMemoryRepository
+import com.netflix.spinnaker.keel.test.resource
+import com.netflix.spinnaker.time.MutableClock
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import io.mockk.clearAllMocks
+import java.time.Duration
+import java.time.Instant
+import java.time.ZoneId
+import strikt.api.expect
+import strikt.api.expectThat
+import strikt.assertions.containsExactly
+import strikt.assertions.isEmpty
+import strikt.assertions.isEqualTo
+import strikt.assertions.isNotEmpty
+import strikt.assertions.isNotNull
+
+class ApplicationServiceTests : JUnit5Minutests {
+  class Fixture {
+    val clock: MutableClock = MutableClock(
+      Instant.parse("2020-03-25T00:00:00.00Z"),
+      ZoneId.of("UTC")
+    )
+    val deliveryConfigRepository = InMemoryDeliveryConfigRepository(clock)
+    val artifactRepository = InMemoryArtifactRepository(clock)
+    val resourceRepository = InMemoryResourceRepository(clock)
+    val repository: KeelRepository = combinedInMemoryRepository(
+      deliveryConfigRepository = deliveryConfigRepository,
+      artifactRepository = artifactRepository,
+      resourceRepository = resourceRepository,
+      clock = clock
+    )
+    val applicationService = ApplicationService(repository)
+
+    val application = "fnord"
+
+    val artifact = DebianArtifact(
+      name = application,
+      deliveryConfigName = "manifest",
+      reference = "fnord",
+      statuses = setOf(ArtifactStatus.RELEASE)
+    )
+
+    val clusterDefaults = ClusterSpec.ServerGroupSpec(
+      launchConfiguration = ClusterSpec.LaunchConfigurationSpec(
+        instanceType = "m4.2xlarge",
+        ebsOptimized = true,
+        iamRole = "fnordInstanceProfile",
+        keyPair = "fnordKeyPair"
+      )
+    )
+
+    val environments = listOf("test", "staging", "production").associateWith { name ->
+      Environment(
+        name = name,
+        constraints = if (name == "production") {
+          setOf(
+            DependsOnConstraint("staging"),
+            ManualJudgementConstraint(),
+            PipelineConstraint(pipelineId = "fakePipeline")
+          )
+        } else {
+          emptySet()
+        },
+        resources = setOf(
+          // cluster with new-style artifact reference
```

Comment:
The "old-style" cluster spec with image details that existed in the original tests was important to check that retrieving cluster -> artifact linkage worked for that case too.

Reply:
ah ok, i'll add it back in